# DNAmethylation_ArrayAnalysis

## Objective
This educational module demonstrates the steps needed to process raw files generated by Illumina's Infinium beadchip array 450k (or the 850k EPIC chip) and perform differential methylation analysis at the probe level. The workflow uses R and Bioconductor packages and includes downloading raw data files from the Gene Expression Omnibus (GEO) repository. Both R code and the same R code within a Jupyter Notebook environment are available.

## Introduction
Epigenetic alteration refers to functionally relevant changes to the genome that do not involve a change in the nucleotide sequence but alter how genes are expressed. Examples of mechanisms that produce such changes are DNA methylation and histone modification. DNA methylation is a process by which methyl groups are
added to the cytosine molecule of cytosine-guanine dinucleotides (CpG) of the DNA. Methylation can change the activity of a DNA segment without changing the sequence. When located in a gene promoter, DNA methylation typically acts to repress gene transcription. DNA methylation is essential for normal development
and is associated with a number of key processes including genomic imprinting, X-chromosome inactivation, repression of transposable elements, aging, and carcinogenesis.

Illumina’s Infinium BeadChips have provided a popular, user-friendly, and affordable alternative to whole-genome bisulfate sequencing (WGBS). Like WGBS this technology is based on sodium bisulphite conversion of DNA, but with subsequent single base resolution genotyping of targeted CpG sites using probes on a microarray. For each methylation locus, two signals of interest are recorded: One signal measures the amount of methylated DNA (Meth), and the other measures the amount of unmethylated DNA (Unmeth). The proportion Meth/(Unmeth+Meth) is the methylation ratio (referred to as β) in the population of cells from which the DNA was extracted. This proportion is used for downstream analysis.

## Setup and Installation
The module was conducted using R version 4.4.2 in a Jupyter Notebook with R kernel. The associated Anaconda environment for the Jupyter Notebook is provided in file 'environment_R_4.4.2.yaml'. If you run the R code, ensure the proper version is installed on your machine along with the required libraries. To install Bioconductor packages, packge **BiocManager** is required. You can install it using the following line:

```R
if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
```

Then use package BiocManager to install the Bioconductor packages needed in this module:

```R
if (!requireNamespace("GEOquery", quietly = TRUE)) BiocManager::install("GEOquery")
if (!requireNamespace("minfi", quietly = TRUE)) BiocManager::install("minfi")
if (!requireNamespace("IlluminaHumanMethylation450kmanifest", quietly = TRUE)) BiocManager::install("IlluminaHumanMethylation450kmanifest")
if (!requireNamespace("IlluminaHumanMethylation450kanno.ilmn12.hg19", quietly = TRUE)) BiocManager::install("IlluminaHumanMethylation450kanno.ilmn12.hg19")
if (!requireNamespace("limma", quietly = TRUE)) BiocManager::install("limma")
```

If you prefer to work with Jupyter Notebook, import the yaml environment file into your Anaconda, install R and the needed packages as shown above. Alternatively, you can create your own conda environment. For example:
```
conda create -n R_4.4.2
```
Then activate the environment and install the desired version of R:
```
conda activate R_4.4.2
conda install -c conda-forge r-base=4.4.2 r-essentials -y 
```
To install the R kernel, you can start R in a terminal and use install.packages:
```
conda activate R_4.4.2
R
```
```R
> install.packages("IRkernel")
```

If installing any package directly from R using install.packages("package_name") proves to be problematic due to dependencies that require a specific package version, you may install such packages using conda after activating your environment. For example:
```
conda activate r_env
conda install -c r r-Matrix=7.3-64
conda install -c r r-matrixStats=1.5.0
conda install -c r r-locfit=1.5-9.12
```

